src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,263 @@
+     */
+    function _getTotalOperatorsForQuorumAtBlockNumber(uint8 quorumNumber, uint32 blockNumber) internal view returns (uint32){
+        // store list length in memory
+        uint256 totalOperatorsHistoryLength = _totalOperatorsHistory[quorumNumber].length;
+        // if there are no entries in the total operator history, return 0
+        if (totalOperatorsHistoryLength == 0) {
+            return 0;
+        }
+
+        // if `blockNumber` is from before the `quorumNumber` existed, return `0`
+        if (blockNumber < _totalOperatorsHistory[quorumNumber][0].fromBlockNumber) {
+            return 0;
+        }
+
+        // loop backwards through the total operator history to find the total number of operators at the given block number
+        for (uint256 i = 0; i <= totalOperatorsHistoryLength - 1; i++) {
+            uint256 listIndex = (totalOperatorsHistoryLength - 1) - i;
+            OperatorIndexUpdate memory totalOperatorUpdate = _totalOperatorsHistory[quorumNumber][listIndex];
+            // look for the first update that began before or at `blockNumber`
+            if (totalOperatorUpdate.fromBlockNumber <= blockNumber) {
+                return _totalOperatorsHistory[quorumNumber][listIndex].index;
+            }
+        }        
+        return _totalOperatorsHistory[quorumNumber][0].index;
+    }
+    
+
+    /// @notice Returns the index of the `operatorId` at the given `blockNumber` for the given `quorumNumber`, or `OPERATOR_DEREGISTERED_INDEX` if the operator was not registered for the `quorumNumber` at blockNumber
+    function _getIndexOfOperatorForQuorumAtBlockNumber(bytes32 operatorId, uint8 quorumNumber, uint32 blockNumber) internal view returns(uint32) {
+        uint256 operatorIndexHistoryLength = _operatorIdToIndexHistory[operatorId][quorumNumber].length;
+        // loop backward through index history to find the index of the operator at the given block number
+        for (uint i = 0; i < _operatorIdToIndexHistory[operatorId][quorumNumber].length; i++) {
+            uint256 listIndex = (operatorIndexHistoryLength - 1) - i;
+            OperatorIndexUpdate memory operatorIndexUpdate = _operatorIdToIndexHistory[operatorId][quorumNumber][listIndex];
+            if (operatorIndexUpdate.fromBlockNumber <= blockNumber) {
+                // one special case is that this will be OPERATOR_DEREGISTERED_INDEX if the operator was deregistered from the quorum
+                return operatorIndexUpdate.index;
+            }
+        }
+
+        // the operator is still active or not in the quorum, so we return the latest index or the default max uint32
+        // this will be hit if `blockNumber` is before when the operator registered or the operator has never registered for the given quorum
+        return OPERATOR_DEREGISTERED_INDEX;
+    }
+}
```

Comment:
noticed there's no storage gap variable for this contract. looks like it's non-upgradeable right now, but I believe it's supposed to be upgradeable

Reply:
checked with @gpsanant -- sounds like the intention was to make this contract (IndexRegistry) upgradeable, so we should add that functionality (inherit from Initializable, add initializer as necessary, add storage gap, disable initializers in constructor)